server/supabase/migrations/20240609010604_helper_party_api_keys.sql

```diff
@@ -0,0 +1,13 @@
+create table
+helper_party_api_keys (
+uuid uuid default gen_random_uuid() primary key,
+helper_party_uuid uuid references helper_parties not null,
+hashed_api_key varchar(255) not null,
+created_at timestamp default current_timestamp not null,
+expires_at timestamp default current_timestamp + interval '1 year' not null,
+revoked boolean default false not null
```

Comment:
can you elaborate why we need an extra flag here? One may think that to revoke a key, you can set `expires_at` in the past and wait until this change is propagated

Reply:
Main reason is so that we retain more information, e.g., revoking a key is different than it expiring, and we'd want to provide different error messages based on those different facts.

Reply:
My main concern is the validity check - you must not forget to check both and if `expires_at` is still in the future and revoked is set, that looks weird. Maybe, as an alternative, you could use `status` or `modify_reason`/`modified_at`? Validity check can only look at `expires_at` and that is simpler

Reply:
If we have a `status` column, then it becomes a bit less clear to a future developer that they need to do a validity check on `expires_at` and not `status`, and similarly with `modify_reason`/`modify_at` you need to assure that it also changes expiry. 

Here's another idea. It would require slightly more work, but would be easier to worry about. Create two tables:

`helper_party_api_key_creation_events` and `helper_party_api_key_revocation_events`. The first table would have everything in this PR except the `revoked` column, and the second table would have: `uuid, 
helper_party_api_key_creation_event_uuid (join column), helper_party_uuid, created_at`.

Then, we'd create a view (possibly materialized), which would be used for the validity checks
```
select 
    c.helper_party_uuid, 
  , c.hashed_api_key
  , c.expires_at < current_timestamp() and r.uuid is null as valid
  , case 
      when r.uuid is not null as 'revoked'
      when c.expires_at >= current_timestamp() as 'expired'
      else null
from 
  helper_party_api_key_creation_events c
left outer join
  helper_party_api_key_revocation_events r
on c.uuid = r.helper_party_api_key_creation_event_uuid
```

This does a few nice things:
1. Maintains all information (e.g., nothing is overwritten or modified, just computed as a result of events.)
2. Minimizes the risk of incorrect validation, as stored directly on the table.
3. Minimized the risk of incorrectly invalidating, as a future developer just needs to create that event, rather than make sure to properly modify the `helper_party_api_key` table.

The only potential downside is that it moves some of business logic into the database. For this, it seems like there are essentially 3 places where logic can live (and things can go wrong):
1. When events are created or modified (revoked). In your example, this would require making sure that when a revoke action happens, the developer needs to correctly update the expiration.
2. When api keys are validated. This requires potentially checking expires_at and revoked.
3. In the database, as a view computed over actions.

After writing this all out, I think I like 3/ the best. Also, this type of event driven architecture is a possible solution to the problem you mention about state management across between the management plane and multiple control planes.

Reply:
With the event sourcing, you typically have only one timestamp which is the timestamp of the event. Then you can build a snapshot and/or the latest status by applying these events according to their timestamp and looking at the resulting object. So you technically can have only one table that records two event types - create/revoke and if revoke is the last event, then key is no longer valid.

This could be more powerful than what we want here - I was thinking that `expiry_at` + `modify_reason = "REVOKE"` is enough information for an engineer to see what is going on, even if reason is unstructured. I don't want to overcompicate things here, so if you are not inclined with this, I think the current proposal is better than even sourcing because it is less complex